### PR TITLE
Ultralytics/YOLOv3 DeepSparse Examples

### PR DESCRIPTION
**Pull request by bfineran**

### CODE REVIEW

File: examples/ultralytics-yolov3/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Ultralytics/yolov3-DeepSparse Inference Examples
+This directory contains examples of benchmarking, annotating, and serving inferences
+of YOLOv3 models from the [ultralytics/yolov3](https://github.com/ultralytics/yolov5)
+repository using the DeepSparse engine. The DeepSparse engine is able to achieve
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/README.md:20]:
```suggestion
repository using the DeepSparse Engine. The DeepSparse Engine is able to achieve
```

File: examples/ultralytics-yolov3/README.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    --quantized-inputs
+```
+
+Note for quantized performance, your CPU must support VNNI instructions.
+
+
+
+# Example YOLO DeepSparse Flask Server
+
+To illustrate how the DeepSparse engine can be used for YOLO model deployments, this directory
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/README.md:71]:
```suggestion
To illustrate how the DeepSparse Engine can be used for YOLO model deployments, this directory
```

File: examples/ultralytics-yolov3/README.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    --quantized-inputs
+```
+
+You can leave that running as a detached process or in a spare terminal.
+
+This starts a Flask app with the DeepSparse Engine as the inference backend, accessible at `http://0.0.0.0:5543` by default.
+
+The app exposes HTTP endpoints at:
+- `/info` to get information about the compiled model
+- `/predict` to send images to the model and receive detected in response.
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/README.md:96]:
```suggestion
- `/predict` to send images to the model and receive as detected in response.
```

File: examples/ultralytics-yolov3/README.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    The number of images should match the compiled model's batch size.
+
+For a full list of options, run `python server.py -h`.
+
+Currently, the server is set to do pre-processing for the yolov3-spp
+model, if other models are used, the image shape, output shapes, and
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/README.md:102]:
```suggestion
model; if other models are used, the image shape, output shapes, and
```

File: examples/ultralytics-yolov3/README.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+anchor grids should be updated. 
+
+### Client
+
+`client.py` provides a `YoloDetectionClient` object to make requests to the server easy.
+The file is self documented.  See example usage below:
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/README.md:108]:
```suggestion
The file is self-documented.  See example usage below:
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<sub>&copy; 2021-present Neuralmagic, Inc. // [Neural Magic Legal](https://neuralmagic.com/legal)</sub> \n",
+    "\n",
+    "# Running YOLO Models on the DeepSparse Engine\n",
+    "\n",
+    "In [this blog post](https://neuralmagic.com/blog/benchmark-yolov3-on-cpus-with-deepsparse/) the Neural Magic team describes how achieving real time inference speeds on [Ultralytic's YOLOv3](https://github.com/ultralytics/yolov5) using CPUs is possible. This notebook provides a step-by-step walkthrough for using the [DeepSparse](https://github.com/neuralmagic/deepsparse) inference engine to attain and benchmark these speeds.\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:11]:
```suggestion
    "In [this blog post](https://neuralmagic.com/blog/benchmark-yolov3-on-cpus-with-deepsparse/) the Neural Magic team describes how achieving real-time inference speeds on [Ultralytic's YOLOv3](https://github.com/ultralytics/yolov5) using CPUs is possible. This notebook provides a step-by-step walkthrough for using the [DeepSparse](https://github.com/neuralmagic/deepsparse) Engine to attain and benchmark these speeds.\n",
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+    "\n",
+    "This notebook runs pre-sparsified models downloaded from the [SparseZoo](https://github.com/neuralmagic/sparsezoo). For information on creating custom sparsified YOLO models with [SparseML](https://github.com/neuralmagic/sparseml), check out the [SparseML-Ultralytics integration documentation](https://github.com/neuralmagic/sparseml/blob/main/integrations/ultralytics/README.md).\n",
+    "\n",
+    "\n",
+    "In this notebook, you will:\n",
+    "- Compile a model with the DeepSparse engine\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:17]:
```suggestion
    "- Compile a model with the DeepSparse Engine\n",
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+    "- Use the DeepSparse engine to run a sample inferences with YOLOv3 and annotate the output\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:18]:
```suggestion
    "- Use the DeepSparse Engine to run a sample inference with YOLOv3 and annotate the output\n",
```

Comment (jeanniefinks) [reply examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:18]: will the user be running more than one inference or just one reference? my suggested edit assumes one. please update though if it's more than one

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    "- Run benchmarks of YOLOv3 on the DeepSparse engine\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:19]:
```suggestion
    "- Run benchmarks of YOLOv3 on the DeepSparse Engine\n",
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## SparseZoo Stubs\n",
+    "[SparseZoo](https://github.com/neuralmagic/sparsezoo) stubs refer to models and SparseML recipes that are stored in the SparseZoo. They can be used to easily access these files through the `sparsezoo.Zoo` helper class, or even to load these models directly into DeepSparse.\n",
+    "\n",
+    "The following cell defines SparseZoo stubs for 3 different models:\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:46]:
```suggestion
    "The following cell defines SparseZoo stubs for three different models:\n",
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+    "PRUNED_STUB = \"zoo:cv/detection/yolo_v3-spp/pytorch/ultralytics/coco/pruned-aggressive_97\"\n",
+    "PRUNED_QUANT_STUB = \"zoo:cv/detection/yolo_v3-spp/pytorch/ultralytics/coco/pruned_quant-aggressive_94\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Inference\n",
+    "In the following cells, you will download some sample images from the SparseZoo, compile a model with the DeepSparse engine, and use this model to perform a sample inference.\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:70]:
```suggestion
    "In the following cells, you will download some sample images from the SparseZoo, compile a model with the DeepSparse Engine, and use this model to perform a sample inference.\n",
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+    "\n",
+    "The basic flow for using the DeepSparse engine for inference is just two steps: compiling a model and running a forward pass with the model:\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:72]:
```suggestion
    "The basic flow for using the DeepSparse Engine for inference is just two steps: compiling a model and running a forward pass with the model:\n",
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+    "outputs = model([numpy.random.randn(1, 3, 640, 640)])\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Download Sample Images\n",
+    "The following cell uses a helper function to download sample images from the SparseZoo"
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:91]:
```suggestion
    "The following cell uses a helper function to download sample images from the SparseZoo:"
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Benchmark\n",
+    "The following cells run benchmarks of the base, pruned, and pruned-quantized models.  Benchmark times are end-to-end and include pre-processing, inference, post-processing and NMS.  The defualt setting is a \"real-time\" batch-size 1 example with the maximum number of CPU cores available used. To try benchmarking other deployment scenarios, edit the `BATCH_SIZE`, `NUM_CORES`, and `IMAGE_SIZE` variables.\n",
+    "\n",
+    "The benchmarking function provided below calls into the `benchmark.py` script that is also provided in this examples directory.  For best results, that script should be used as processes in the jupyter notebook may interfere with the benchmark."
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:185]:
```suggestion
    "The benchmarking function provided below calls into the `benchmark.py` script that is also provided in this examples directory.  For best results, that script should be used as processes in the Jupyter Notebook may interfere with the benchmark."
```

Comment (jeanniefinks) [reply examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:185]: @bfineran There's a word or two missing in that last sentence. e.g.,  should it read "as they may interfere"?

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+    "benchmark(PRUNED_QUANT_STUB, quantized=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Next Steps\n",
+    "\n",
+    "Congratulations, you have created run a sample inference of YOLOv3 using the DeepSparse engine and benchmarked its performance across various deployment settings. Next steps can include:\n",
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:273]:
```suggestion
    "Congratulations, you have created run a sample inference of YOLOv3 using the DeepSparse Engine and benchmarked its performance across various deployment settings. Next steps can include:\n",
```

File: examples/ultralytics-yolov3/deepsparse-yolov3.ipynb [hunk 16]
```diff
@@ -0,0 +1,308 @@
...
+    "* Sparsify other models, or this model for other datasets using the SparseML-Ultralytics integration\n",
+    "* Try the other exmaples in this directory\n",
+    "* Learn more about [Neural Magic](https://github.com/neuralmagic) and [Ultralytic's](https://github.com/ultralytics) open sourced repositories."
```
Comment (jeanniefinks) [inline examples/ultralytics-yolov3/deepsparse-yolov3.ipynb:276]:
```suggestion
    "* Learn more about [Neural Magic](https://github.com/neuralmagic) and [Ultralytic's](https://github.com/ultralytics) open-sourced repositories."
```

**Review by markurtz — APPROVED** [review]

**Comment by markurtz** [comment]
@jeanniefinks committed your suggestions

**Review by jeanniefinks — APPROVED** [review]

**Pull request merged**